### PR TITLE
feat(*): MapDeclaration Parse & Print

### PR DESCRIPTION
**Pull request by Jonathan-Casey**
# Closes #<CORRESPONDING ISSUE NUMBER>

The implementation of the map declaration feature will be delivered incrementally through a phased approach. The first phase involves updating the parser grammar, & enabling both Parse and Print for map types. As part of this process, test data for both the cto and ast representation of a map declaration is added to test the print and parse functionalities.

At its core, a `MapDeclaration` consists of one `MapKeyType` and a general `AggregateValueType`, which may in some cases be specialized to an `AggregateRelationshipValueType`—both of which may be annotated, as too can the parent declaration.

For more information on the design specification see [here](https://github.com/accordproject/concerto/wiki/Aggregate-Types-Design)

### Changes

- <ONE> Updated Parser.pegjs
- <TWO> Updated Printer
- <THREE> Added test data for CTO and AST representation 

### Related Issues
- Issue #447 

### Author Checklist
- [X] Ensure you provide a [DCO sign-off](https://github.com/probot/dco#how-it-works) for your commits using the `--signoff` option of git commit.
- [X] Vital features and changes captured in unit and/or integration tests
- [X] Commits messages follow [AP format](https://github.com/accordproject/techdocs/blob/master/DEVELOPERS.md#commit-message-format)
- [ ] Extend the documentation, if necessary
- [ ] Merging to `main` from `fork:branchname`

### CODE REVIEW

File: packages/concerto-core/types/lib/model/identifiable.d.ts
Comment (mttrbrts) [inline packages/concerto-core/types/lib/model/identifiable.d.ts:1]: This change is unrelated I believe, likely caused by me not building types in a previous PR.

File: packages/concerto-core/types/lib/modelutil.d.ts
Comment (mttrbrts) [inline packages/concerto-core/types/lib/modelutil.d.ts:1]: Unrelated, as above

File: packages/concerto-core/types/lib/serializer/jsongenerator.d.ts
Comment (mttrbrts) [inline packages/concerto-core/types/lib/serializer/jsongenerator.d.ts:1]: Unrelated, as above

File: packages/concerto-core/types/lib/serializer/jsonpopulator.d.ts
Comment (mttrbrts) [inline packages/concerto-core/types/lib/serializer/jsonpopulator.d.ts:1]: Unrelated, as above

File: packages/concerto-core/types/lib/serializer/resourcevalidator.d.ts
Comment (mttrbrts) [inline packages/concerto-core/types/lib/serializer/resourcevalidator.d.ts:1]: Unrelated, as above

File: packages/concerto-cto/lib/parser.pegjs [hunk 1]
```diff
@@ -1565,6 +1619,8 @@ SourceElement
   / EnumDeclaration
   / ConceptDeclaration
   / ScalarDeclaration
-
+  / MapDeclaration
+  / MapPropertyKeyDeclaration
```
Comment (mttrbrts) [inline packages/concerto-cto/lib/parser.pegjs:1623]: I think that this means that I could declare an orphaned Key or Value without an enclosing MapDeclaration?

File: packages/concerto-cto/lib/printer.js [hunk 2]
```diff
@@ -197,28 +197,28 @@ function modifiersFromMetaModel(mm){
...
+function propertyFromMetaModel(prop) {
     let result = '';
 
-    if (mm.decorators) {
-        result += decoratorsFromMetaModel(mm.decorators, '  ');
+    if (prop.decorators) {
+        result += decoratorsFromMetaModel(prop.decorators, '  ');
     }
-    if (mm.$class === `${MetaModelNamespace}.RelationshipProperty`) {
+    if (prop.$class === `${MetaModelNamespace}.RelationshipProperty` || prop.isRelationship) {
```
Comment (mttrbrts) [inline packages/concerto-cto/lib/printer.js:210]: This suggests to me that the metamodel is wrong.

Did you consider boxing a relationship to a Map inside a RelationshipProperty?

As it is, I believe that this will give us two ways to represent relationships to things.
1. Using a `RelationshipProperty`
2. Using a `MapDeclaration` with `isRelationship` set to `true`

File: packages/concerto-cto/test/cto/map.cto
Comment (mttrbrts) [inline packages/concerto-cto/test/cto/map.cto:1]: Can you add some negative tests, tests that include decorators, and tests with more variation in whitespace, please?

**Review by mttrbrts — CHANGES_REQUESTED** [review]

File: packages/concerto-core/types/lib/model/identifiable.d.ts
Comment (Jonathan-Casey) [reply packages/concerto-core/types/lib/model/identifiable.d.ts:1]: yes, can ignore this.

File: packages/concerto-cto/lib/parser.pegjs (hunk 1 above)
Comment (Jonathan-Casey) [reply packages/concerto-cto/lib/parser.pegjs]: You are correct

File: packages/concerto-cto/test/cto/map.cto
Comment (Jonathan-Casey) [reply packages/concerto-cto/test/cto/map.cto:1]: Negative case and decorator coverage added. Does it make sense to add coverage for whitespace, given the parser explicitly ignores whitespace between the other tokens:

```
WhiteSpace "whitespace"
  = "\t"
  / "\v"
  / "\f"
  / " "
  / "\u00A0"
  / "\uFEFF"
  / Zs
  
/* Skipped */

__
  = (WhiteSpace / LineTerminatorSequence / Comment)*

_
  = (WhiteSpace / MultiLineCommentNoLineTerminator)*
```

File: packages/concerto-cto/lib/printer.js (hunk 2 above)
Comment (Jonathan-Casey) [reply packages/concerto-cto/lib/printer.js:210]: On rereading your comment it sounds like you mean something like:

```
concept RelationshipProperty extends Property {
  o TypeIdentifier type
  o  MapValue 
}
```

which still inherits the design issues pointed out above.

I'm curious why you think the initial solution may be wrong based on the branch

`    if (prop.$class === `${MetaModelNamespace}.RelationshipProperty` || prop.isRelationship) {
`
Why does this suggest the metamodel is incorrect?

Comment (Jonathan-Casey) [reply packages/concerto-cto/lib/printer.js]: @mttrbrts changes added based on prior discussion. You can find the updated metamodel for reference [here](https://github.com/accordproject/concerto-metamodel/pull/2)

**Review by mttrbrts — APPROVED** [review]
